### ISSUE TITLE
Ordering

### ISSUE THREAD

**Issue by Stono** [roboll/helmfile#137]
Firstly, 
Thanks for this project, it's awesome - you've saved me a big headache.

However, quick Q.   I need to control the order in which the charts are deployed.

Is there any way (hacky or not) that I can do that, without using labels and having two helmfile runs?

**Comment by sstarcher** [comment]
not currently

**Comment by mumoshu** [comment]
@Stono Hi! Just curious but why do you want the ordering?

**Comment by Stono** [comment]
Hello @mumoshu!
Great question, let me try and explain...

We have a helmfile which can take an empty kubernetes cluster and turn it into what we call the "delivery platform", however the components need to be deployed in a certain order, for a variety of reasons:

  - We need to deploy a chart which contains CustomResourceDefinitions **before** we deploy a chart which creates instances of those CustomResources
  - We use a service mesh, which uses a MutatingWebhookHandler in order to inject the sidecar into Pods as they're deployed.  This needs to be configured before any Deployments, ReplicaSets or Pods are deployed.

As a result, we've ended up with a bit of a hacky script that looks like this:

```
echo "Creating MutatingWebhookHandler certificates..."
cd manual
./install-sidecar-injection.sh
cd ..

echo "Deploying istio (first phase)..."
helmfile --selector component=istio,order=1 sync --args="--wait --timeout=300"

echo "Adding the MutatingWebhookHandler..."
kubectl apply -f manual/istio-sidecar-injector-with-ca-bundle.yaml

echo "Deploying istio (second phase)..."
helmfile --selector component=istio,order!=1 sync --args="--wait --timeout=300"

echo "Deploying everything else..."
helmfile --selector component!=istio sync --args="--wait --timeout=300"
```

<img width="969" alt="screen shot 2018-05-11 at 09 22 26" src="https://user-images.githubusercontent.com/1486729/39914585-5c385656-54fd-11e8-83b6-7c9a776a11d3.png">

**Comment by mumoshu** [comment]
@Stono Hi. Thank you so much for the detailed explanation!
Makes sense.

Btw, would this suggested feature https://github.com/roboll/helmfile/issues/151#issuecomment-391165364 help your use-case, too?

**Comment by Stono** [comment]
Yes it would so long as the files were processed in order 👌

**Comment by gtaylor** [comment]
To throw in another usage case, we have Helm charts for Roles and RoleBindings. IIRC, the latter will fail if you try to bind a Role that doesn't exist yet.

We also Chart StorageClasses, which have to be the very first thing provisioned (our default storage class, in particular).

**Comment by mumoshu** [comment]
I have a WIP PR for #151 with lexical ordering at #160. Would you mind taking a look?

**Comment by sstarcher** [comment]
@gtaylor what's the logic for having roles and rolebindings in separate charts and not in the same chart?

**Comment by gtaylor** [comment]
You don't always need both.

**Comment by sstarcher** [comment]
I don't understand why you would have roles, but not bind them to anything.  Seems like they go together to me and I would also enable/disable them through flags instead of having them in separate charts, but I don't have your full use-case.

**Comment by mumoshu** [comment]
#247 would allow more granular control over ordering.

**Comment by mumoshu** [comment]
@Stono Can we close this as resolved via #160 #266, and #339?

**Comment by mumoshu** [comment]
As answered in #591, I think sub-helmfiles can be used for ordering. It was known to be very repetitive, but after #587 the repetition is greatly reduced.

Closing this as resolved, but please feel free to reopen or create an another issue if you have any other request! Thanks.

**Comment by borgilb** [comment]
@mumoshu looked at the docs but not 100% sure how ordering works. 

Can I just do this and they will be installed in the order that they are specified?
```
releases:
- name: chart-that-installs-a-crd
  chart: stable/chart-that-installs-a-crd
- name: chart-that-uses-that-crd
  chart: chart-that-uses-that-crd
```

**Comment by mumoshu** [comment]
Nope. Releases are handled concurrently per `--concurrency N` where `N` is the number of releases in each helmfile.yaml :)

**Comment by andrewnazarov** [comment]
Thus basically you can achieve some ordering by setting concurrency to 1 :) But probably it's better to do something like this:

```
helmfiles:
- install-crd.yaml

releases:
- name: chart-that-uses-that-crd
   chart: chart-that-uses-that-crd
```

**Comment by mumoshu** [comment]
@andrewnazarov Thanks for the comment! Yeah I think that's correct.

Also see #715 that will give you concurrency AND ordering without splitting helmfile.yaml files.

**Comment by michaelruigrok** [comment]
For those here today, you can use `needs` to specify required order/dependencies
https://helmfile.readthedocs.io/en/latest/#dag-aware-installationdeletion-ordering-with-needs
